src/Analyser/NodeScopeResolver.php

```diff
@@ -1843,75 +1842,72 @@ function (MutatingScope $scope) use ($expr, $nodeCallback, $context): Expression
 				&& in_array($functionReflection->getName(), ['array_push', 'array_unshift'], true)
 				&& count($expr->getArgs()) >= 2
 			) {
-				$argumentTypes = [];
-				foreach (array_slice($expr->getArgs(), 1) as $callArg) {
-					$callArgType = $scope->getType($callArg->value);
-					if ($callArg->unpack) {
-						$iterableValueType = $callArgType->getIterableValueType();
-						if ($iterableValueType instanceof UnionType) {
-							foreach ($iterableValueType->getTypes() as $innerType) {
-								$argumentTypes[] = $innerType;
+				$arrayArg = $expr->getArgs()[0]->value;
+				$arrayType = $scope->getType($arrayArg);
+				$callArgs = array_slice($expr->getArgs(), 1);
+
+				/**
+				 * @param Arg[] $callArgs
+				 * @param callable(?Type, Type, bool=): void $setOffsetValueType
+				 */
+				$setOffsetValueTypes = static function (Scope $scope, array $callArgs, callable $setOffsetValueType): void {
+					foreach ($callArgs as $callArg) {
+						$callArgType = $scope->getType($callArg->value);
+						if ($callArg->unpack) {
+							if ($callArgType->isIterableAtLeastOnce()->no()) {
+								continue;
 							}
-						} else {
-							$argumentTypes[] = $iterableValueType;
+							$iterableValueType = $callArgType->getIterableValueType();
+							$isOptional = !$callArgType->isIterableAtLeastOnce()->yes();
+							if ($iterableValueType instanceof UnionType) {
```

Comment:
Not sure if this is even needed, I think the UnionType should handle it. Will add a test and check later

Reply:
On second thought: yes ,the constant array needs this :)